connectors/sink/src/main/scala/kafka/connect/http/sink/HttpWriter.scala

```diff
@@ -16,54 +16,97 @@
+import scala.concurrent.duration.{Duration, DurationLong}
 
 class HttpWriter(
     private val config: HttpSinkConfig,
     private val dispatcher: Dispatcher,
     private val formatter: Formatter,
-    private val authenticator: Option[Authenticator]
-) extends Logging {
+    private val authenticator: Option[Authenticator],
+    private val context: Option[SinkTaskContext]
+)(implicit ec: ExecutionContext)
+    extends Logging {
 
-  var batches: List[SinkRecord] = List()
-  var nextCheck: Instant        = Instant.MIN
+  var batches: List[SinkRecord]                                     = List()
+  val lastCommitted: mutable.Map[TopicPartition, OffsetAndMetadata] = mutable.Map.empty
 
   def put(records: List[SinkRecord]): Unit = {
     batches = batches ++ records
     // send only one batch in batches
-    if (batches.size > 0 && nextCheck.isBefore(Instant.now())) {
+    if (batches.size >= config.batchSize || lastCommitted.isEmpty) {
       val (toSend, remaining) = batches.splitAt(config.batchSize)
-      sendBatch(toSend, true)
-      nextCheck = Instant.now().plus(config.maxTimeout, ChronoUnit.MILLIS)
+      sendBatch(toSend, lastCommitted.isEmpty)
       batches = remaining
     }
   }
 
-  def flush(): Unit = {
-    batches.grouped(config.batchSize).foreach(sendBatch(_, nextCheck.isBefore(Instant.now())))
+  def flush(): Map[TopicPartition, OffsetAndMetadata] = {
+    batches.grouped(config.batchSize).foreach(sendBatch(_, lastCommitted.isEmpty))
     batches = List()
+    lastCommitted.toMap
   }
 
   private def sendBatch(records: List[SinkRecord], failFast: Boolean): Unit = {
     val body = formatter.toOutputFormat(records)
     val headers =
       authenticator.fold(config.httpHeaders)(a => config.httpHeaders + (config.authHeaderName -> a.authHeader()))
-    dispatcher.send(headers, body, failFast)
+    dispatcher.send(headers, body, failFast) match {
+      case None =>
+        records.foreach(r => updateLastCommitted(toTopicPartition(r), r.kafkaOffset()))
+      // send failed => we are not going to commit and instead of that we'll be pausing consumption
+      case Some(e) => pauseFor(e, records.map(toTopicPartition), dispatcher.pauseTime)
+    }
+  }
+
+  /** Pauses consumption of topic partitions for a given duration
+    * @param e error to verify that maximum amount of retries hasn't been reached.
+    * @param tp list of partitions
+    * @param duration duration of pause
+    */
+  private def pauseFor(e: SinkError, tp: List[TopicPartition], duration: Duration): Unit =
+    e match {
+      case er: MaxAmountOfRetriesReached => throw er
+      case _ =>
+        context.foreach(_.pause(tp: _*))
+        logger.info(s"Pausing partitions ${tp.mkString(",")} for ${duration}")
+        unpause(tp, duration).onComplete(_ => logger.info(s"Partitions un-paused => ${tp.mkString(",")}"))
+    }
+
+  private def unpause(tp: List[TopicPartition], duration: Duration) = Future {
+    Thread.sleep(duration.toMillis)
```

Comment:
Moving the sleep to the unpause function will not prevent blocking the main thread and being unresponsive to connector restart requests?

Reply:
Exactly. This `Future` will run with a separate execution context/different thread pool, so there won't be any blocking at all. In principle it should be sufficient to have a constant size thread pool.